### PR TITLE
On-the-fly endianness conversion

### LINKED ISSUE #124: numba issues with 0.6.1
Hi,

I get some numba issues with 0.6.1 that were not there with 0.6.0.

I am not entirely sure how to make a simple reproduction of the problem. But it appears to happen when I use my cubedsphere package (see cubedsphere.readthedocs.io).

```py
(exorad) schneider@tarf plot % python plot.py
time needed to build regridder: 10.789268970489502
Regridder will use conservative method
Traceback (most recent call last):
  File "/Volumes/EXTERN/Simulations/exorad/paper/paper_runs/plot/plot.py", line 21, in <module>
    ds_reg = regrid()
  File "/Users/schneider/codes/exo/cubedsphere/cubedsphere/regrid.py", line 293, in __call__
    ds[data] = self._regrid_wrapper(interp, **kwargs)
  File "/Users/schneider/codes/exo/cubedsphere/cubedsphere/regrid.py", line 364, in _regrid_wrapper
    data_out += self.regridder[i](ds_in.isel(**{c.FACEDIM: i}), **kwargs)
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/xesmf/frontend.py", line 447, in __call__
    return self.regrid_dataarray(
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/xesmf/frontend.py", line 522, in regrid_dataarray
    dr_out = xr.apply_ufunc(
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/xarray/core/computation.py", line 1174, in apply_ufunc
    return apply_dataarray_vfunc(
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/xarray/core/computation.py", line 293, in apply_dataarray_vfunc
    result_var = func(*data_vars)
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/xarray/core/computation.py", line 742, in apply_variable_ufunc
    result_data = func(*input_data)
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/xesmf/frontend.py", line 469, in _regrid_array
    outdata = apply_weights(weights, indata, shape_in, shape_out)
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/xesmf/smm.py", line 150, in apply_weights
    outdata_flat = weights.dot(indata_flat.T).T
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/sparse/_coo/core.py", line 901, in dot
    return dot(self, other)
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/sparse/_common.py", line 326, in dot
    return tensordot(a, b, axes=(a_axis, b_axis))
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/sparse/_common.py", line 198, in tensordot
    res = _dot(at, bt, return_type)
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/sparse/_common.py", line 481, in _dot
    return _dot_coo_ndarray_type(a.dtype, b.dtype)(
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/numba/core/dispatcher.py", line 420, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/numba/core/dispatcher.py", line 361, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
non-precise type pyobject
During: typing of argument at /Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/sparse/_common.py (1052)

File "../../../../../../../Users/schneider/anaconda3/envs/exorad/lib/python3.9/site-packages/sparse/_common.py", line 1052:
    def _dot_coo_ndarray(coords1, data1, array2, out_shape):  # pragma: no cover
        <source elided>
        """
        out = np.zeros(out_shape, dtype=dtr)
        ^

This error may have been caused by the following argument(s):
- argument 2: Unsupported array dtype: >f8
```

### PR DESCRIPTION
**Pull request by aulemahal**
Fixes #124.

#111 introduced an unforeseen constrain : machine architectures and data dtypes are now restricted by what `numba` supports. `numba` is used by `sparse` under the hood. I have not heard of problems with machine architectures, but issue #124 is due to data coming from a different computer that uses "big-endian" (numpy's ">f8" dtype).

~This introduces on-the-fly byte-swapping, to convert to little-endian. The performance cost is unmeasured, but should be quite low. I find it a bit ugly to perform that conversion in xESMF instead of `sparse` directly, but this is the fastest solution I could think of.~
EDIT: Instead of converting dtypes on-the-fly, I converted the weights from it's `sparse` format to a `scipy` matrix, which doesn't use numba and thus resolves the problem. It also conserves the dtype, which might be wanted.

I used a function from `numba` directly, should I add it as an explicit dependency?

Also, it might not be necessary to raise a warning?

### CODE REVIEW

**Review by huard — APPROVED** [review]

**Comment by raphaeldussin** [comment]
let's not rush to merge this until we figure out the right course of action

**Comment by huard** [comment]
@raphaeldussin I suggest we merge this, as it now includes a fall-back to scipy.

**Comment by raphaeldussin** [comment]
@huard should we figure out the dask graph/performance issue first before we had more commits?

File: xesmf/smm.py [hunk 1]
```diff
@@ -121,6 +122,15 @@ def apply_weights(weights, indata, shape_in, shape_out):
         Extra dimensions are the same as `indata`.
         If input data is C-ordered, output will also be C-ordered.
     """
+    # Limitation from numba : some big-endian dtypes are not supported.
+    try:
```
Comment (raphaeldussin) [inline xesmf/smm.py:126]: how costly is this test? wouldn't it be cheaper to just test with:

```python
if indata.dtype.byteorder == '>'
```

Comment (aulemahal) [reply xesmf/smm.py:126]: I don't think it's very costly. My intuition comes from the fact that it happens when numba doesn't support the given dtype, so no computation is done at all. But indeed,  it is usually less costly to perform a single check than a try-except call, that's from pure python.

However, I was trying to be as general as I can be here. This change was triggered by numba not supporting a given byte-order, but could there be other things that numba doesn't support, especially on other machines? Evidently, I can't test those with such a computer, but I tried to be ready for other eventualities...

Comment (raphaeldussin) [reply xesmf/smm.py:126]: fair point, I'm ok with sacrificing a bit of performance for compatibility. try/except it is then

**Review by raphaeldussin — APPROVED** [review]

**Pull request merged**